src/Parser/CachedParser.php

```diff
@@ -25,8 +43,22 @@ public function __construct(Parser $originalParser)
 	 */
 	public function parseFile(string $file): array
 	{
+		if ($this->cachedNodesByFileCountMax !== 0 && $this->cachedNodesByFileCount >= $this->cachedNodesByFileCountMax) {
+			$offset = (int) ($this->cachedNodesByFileCountMax / 2);
+
+			$this->cachedNodesByFile = array_slice(
+				$this->cachedNodesByFile,
+				$offset,
+				null,
+				true
+			);
+
+			$this->cachedNodesByFileCount -= $offset;
```

Comment:
Can you explain to me the logic?

Reply:
We remove first half of cached files and keep in amount of cached files (not to call `count` every time we need to parse/get file).

This is materialized count for cached files.